mars/services/web/core.py

```diff
@@ -0,0 +1,175 @@
+
+    def add_instance(self, instance):
+        api_id = id(instance)
+        self._apis[api_id] = instance
+        self.keep_alive_api(api_id)
+        self.remove_expired_api()
+        return api_id
+
+    def get_instance(self, api_id):
+        return self._apis[api_id]
+
+    def keep_alive_api(self, api_id):
+        assert api_id in self._apis, f'API of {api_id} is expired.'
+        self._apis_access_time[api_id] = time.time()
+        # Make order consistent with access time so that iteration can exit early.
+        self._apis_access_time.move_to_end(api_id)
+
+    def remove_expired_api(self):
+        current_time = time.time()
+        expired_api = []
+        for api_id, last_access_time in self._apis_access_time.items():
+            if current_time - last_access_time > self._expired_sec:
+                expired_api.append(api_id)
+            else:
+                # Later last_access_time will be greater so that we can exit early
+                break
+        for api_id in expired_api:
+            self._apis.pop(api_id)
+            self._apis_access_time.pop(api_id)
+
+
+class ServiceWebHandlerBase(MarsRequestHandler):
+    _api_registry = ApiRegistry(_expired_sec)
+    _api_cls = None
+
+    async def post(self, path, **_):
+        api_method_name = path
+        # params format: api_id[None], args, kwargs
+        api_id, args, kwargs = deserialize(self.request.body)
+        try:
+            if not api_id:
+                if hasattr(self, api_method_name):
+                    method = getattr(self, api_method_name)
+                else:
+                    method = getattr(self._api_cls, api_method_name)
+            else:
+                # call method of api instance
+                method = getattr(self._api_registry.get_instance(api_id), api_method_name)
+            if kwargs:
+                # Some methods are decorated with @alru_cache, which doesn't support dict as part of cache key.
+                result = method(*args, **kwargs)
+            else:
+                result = method(*args)
+            if inspect.iscoroutine(result):
+                result = await result
+            self.write(serialize(result))
+        except Exception as e:
+            logger.exception(f'Execute method {api_method_name} with {api_id, args, kwargs} failed, got exception {e}')
+            exc_type, exc_value, exc_traceback = sys.exc_info()
+            self.write(serialize(_HandlerException(exc_type, exc_value, exc_traceback)))
+
+    def _keep_alive_api(self, api_id):
+        self._api_registry.keep_alive_api(api_id)
+        self._api_registry.remove_expired_api()
+
+
+class ServiceWebAPIBase:
+    _service_name = None
+
+    def __init__(self, http_client: AsyncHTTPClient, api_id: int):
+        self._http_client = http_client
+        self._api_id = api_id
+        # TODO(chaokunyang) how to cancel this task
+        self._keep_api_alive_task = asyncio.create_task(self._keep_api_alive())
```

Comment:
By putting keep-alive request in every Web API, unnecessary heart beats will be sent given num of API instances, which is not needed.